### PR TITLE
DList mutation: popFirstOf and popLastOf

### PR DESCRIPTION
**Pull request by Boris-Barboris**
There are currently two ways to remove elements from DList while iterating over it:
* using `Range remove(Range r)` wich takes private DList.Range as a parameter.
* using `Range linearRemove(Take!Range r)` with r being of type Take!(DList.Range).

Both ways are unable to provide an efficient solution to the following problem:
`While iterating the list backwards, remove arbitrary elements.`

Two proposed methods exploit cursor nature of ranges, and provide means to remove elements form the list while iterating over it, in one loop.

Additionaly, I would like to bring up the idea of further encouragement of cursor properties of ranges in Phobos. Quick glance at std.container shows, that many of them share the abovementioned problem, or rely on Take!Range and may or may not be optimal (depends on the compiler) for one-element removals, invoked by take(r, 1).

Please, correct me if I'm wrong.

### CODE REVIEW

File: std/container/dlist.d [hunk 1]
```diff
@@ -630,6 +630,54 @@ Complexity: $(BIGOH 1)
...
+        if (r.empty)
+            return r;
+
+        assert(_root !is null, "Cannot remove from an un-initialized List");
+        assert(r._first, "Remove: Range is empty");
+        BaseNode.connect(r._first._prev, r._first._next);
+        auto after = r._first._next;
+        if (after is _root)
+            return Range(null, null);
+        else
```
Comment (andralex) [inline std/container/dlist.d]: redundant `else`

Comment (andralex) [inline std/container/dlist.d]: Please add a documentation unittest.

File: std/container/dlist.d [hunk 2]
```diff
@@ -630,6 +630,54 @@ Complexity: $(BIGOH 1)
...
+        if (r.empty)
+            return r;
+
+        assert(_root !is null, "Cannot remove from an un-initialized List");
+        assert(r._last, "Remove: Range is empty");
+        BaseNode.connect(r._last._prev, r._last._next);
+        auto before = r._last._prev;
+        if (before is _root)
+            return Range(null, null);
+        else
```
Comment (andralex) [inline std/container/dlist.d]: redundant else - use e.g. `return e ? that : theother;`

File: std/container/dlist.d [hunk 3]
```diff
@@ -816,6 +864,33 @@ private:
 {
     import std.algorithm.comparison : equal;
 
+    auto dl = DList!int([1, 2, 3, 4, 5]);
```
Comment (andralex) [inline std/container/dlist.d]: This is nice, some of it can be moved to the doc unittests.

**Comment by andralex** [comment]
cc @edi33416

**Comment by Boris-Barboris** [comment]
Updated docunittest, comments, made range to shrink as well to keep it in consistent state.

**Pull request merged**